### DISCUSSION TITLE
C# Cannot access a static property of type T in a generic class.

### DISCUSSION THREAD

**Discussion by alexdrl** [dotnet/csharplang#2196] [Q&A] [answered]
Hello, trying to find the solution to a problem I found that I cannot access a static property from a generic method.

The code is like this:
```cs
public class A
{
    public static Guid identifier = new Guid("00000000-0000-0000-0000-000000000000");
}

public class B<T> where T : A
{
    public int Method()
    {
        return T.identifier;
    }
}
```

Why is this not supported? I am missing something?

Found a workaround using custom attributes and reflection, but I think that using reflection for that is suboptimal.

**Comment by EwneoN** [comment]
Try A.identifier

**Comment by qrli** [comment]
Instead of `T.identifier`, just use `A.identifier`.

**Comment by khm1600** [comment]
@EwneoN @qrli 
what if
```cs
    class A
    {
        public static int X => 0;
    }

    class C : A
    {
        public static new int X => 1;
    }

    class B<T> where T : A
    {
        public static int Y => T.X;
    }
```

I'm guessing this is the op's intent.

**Comment by alexdrl** [comment]
@khm1600 Yes, I'm looking for something like an `public static abstract` property.

**Comment by YairHalberstadt** [comment]
Also, take a look at the shapes proposal

**Comment by EwneoN** [comment]
@alexdrl Using a custom attribute and reflection seems like it is possibly your only choice here as "Inheritance in .NET works only on instance base. Static methods are defined on the type level not on the instance level."

https://social.msdn.microsoft.com/Forums/en-US/2e6ce944-f32a-4962-8dbf-861db21cca30/c-can-i-inherit-static-fields-?forum=csharplanguage

If performance is a concern, you could look at in memory caching using a dictionary on a static class to store the Type and PropertyInfo when using reflection to find the fields marked with the custom attribute. Just make sure the cache is thread-safe if you need it to be.

**Comment by qrli** [comment, marked as answer]
> Yes, I'm looking for something like an `public static abstract` property.

There is no such feature in .NET/CLR. `static` means no runtime dispatching. So no inheritance on static members. 

That being said, such feature can be simulated easily with singleton classes as type class. Nothing fancy. That's also how they are implemented in languages supporting that. It is just that: it is a high level feature, while `static` is already bound to the low level foundational feature, which many others rely on.

As an example:
```cs
class ATypeClass {
    public int X { get; set; }
}
class A {
    public static ATypeClass TypeClassInstance = new ATypeClass();
    public ATypeClass TypeClass { get { return TypeClassInstance; } }
}
```

**Comment by MkazemAkhgary** [comment]
I dont feel good about instance variable returning static value.
